core/src/Dscp/Core/Transactions/Types.hs

```diff
@@ -0,0 +1,276 @@
+    , txOuts    :: [TxOut]
+    } deriving (Eq, Ord, Generic, Show)
+
+instance Buildable Tx where
+    build Tx{..} =
+        "Tx { from: " +| txInAcc |+ "; inValue: " +| txInValue |+ "; outs:" +| listF txOuts |+ " }"
+
+type TxId = Hash Tx
+
+data PublicationTx = PublicationTx
+    { ptAuthor    :: Address
+    , ptPrevBlock :: Maybe (Hash PrivateBlock)
+    , ptBlock     :: Hash PrivateBlock
+    } deriving (Eq, Ord, Generic, Show)
+
+instance Buildable PublicationTx where
+    build PublicationTx { ptAuthor, ptPrevBlock, ptBlock } =
+        "Tx { from: " +| ptAuthor |+ "; prev block: " +| ptPrevBlock |+ "; block:" +| ptBlock |+ " }"
+
+type PublicationTxId = Hash PublicationTx
+
+-- | Compute tx id.
+txId :: (Serialise Tx) => Tx -> TxId
+txId = hash
+
+-- | Compute tx id.
+ptxId :: (Serialise PublicationTx) => PublicationTx -> PublicationTxId
+ptxId = hash
+
+-- | Transaction witness. We sign a pair of transaction hash and private
+-- key. The second element is there to authenticate proposed changes
+-- (@kirill.andreev). Public key hash should be equal to the input address.
+-- Also, public key should be the same which used to validate signature.
+data TxWitness = TxWitness
+    { txwSig :: Signature (TxId, PublicKey, ())
+    , txwPk  :: PublicKey
+    } deriving (Eq, Show, Generic)
+
+instance Buildable TxWitness where
+    build TxWitness {..} =
+        "TxWitness { " +| txwSig |+ ", pk: " +| txwPk |+ " }"
+
+-- | Incoming publication message; transaction payload.
+data Publication blockHash addr = Publication
+    { pPrivateBlockHash  :: blockHash
+    , pPreviousBlockHash :: Maybe blockHash
+    }
+    deriving (Eq, Ord, Show, Generic)
+
+data PublicationTxWitness = PublicationTxWitness
+    { pwSig :: Signature (PublicationTxId, PublicKey, Publication (Hash PrivateBlock) Address)
+    , pwPk  :: PublicKey
+    } deriving (Eq, Show, Generic)
+
+instance Buildable PublicationTxWitness where
+    build PublicationTxWitness {..} =
+        "PublicationTxWitness { " +| pwSig |+ ", pk: " +| pwPk |+ " }"
+
+-- | Transaction coupled with witness.
+data TxWitnessed = TxWitnessed
+    { twTx      :: Tx
+    , twWitness :: TxWitness
+    } deriving (Eq, Show, Generic)
+
+instance Buildable TxWitnessed where
+    build TxWitnessed {..} =
+        "TxWitnessed { " +| twTx |+ ", " +| twWitness |+  " }"
+
+-- | Transaction coupled with witness.
+data PublicationTxWitnessed = PublicationTxWitnessed
+    { ptwTx      :: PublicationTx
+    , ptwWitness :: PublicationTxWitness
+    } deriving (Eq, Show, Generic)
+
+instance Buildable PublicationTxWitnessed where
+    build PublicationTxWitnessed {..} =
+        "PublicationTxWitnessed { " +| ptwTx |+ ", " +| ptwWitness |+  " }"
+
+-- | Actual structure in the storage.
+-- |
+-- | For each educator, we'll have a `PublicationOf id ~> LastPublication blockHash`
+-- | (which is effectively `(educator, lastBlockHash)`)
+-- | to determine the lst publication.
+newtype PublicationsOf addr
```

Comment:
I think this should go elsewhere and doesn't belong to the core types.

Reply:
Techically, it is only needed in validator and expander. Where to place it, then?